### PR TITLE
Address #958

### PR DESCRIPTION
**Pull request by miguel-vila**

### CODE REVIEW

**Comment by Baccata** [comment]
Here's what I think we should do : 

1. Introduce a `smithy4s.schema.SchemaType` enumeration (just a good old ADT with only case objects will do)
2. Turn the default method into a `def default[A](id: ShapeId, hints: Hints, schemaType: SchemaType): F[A]`

**Comment by kubukoz** [comment]
I like that.

File: modules/core/src/smithy4s/schema/SchemaVisitor.scala [hunk 1]
```diff
@@ -29,8 +29,8 @@ trait SchemaVisitor[F[_]] extends (Schema ~> F) { self =>
   def union[U](shapeId: ShapeId, hints: Hints, alternatives: Vector[SchemaAlt[U, _]], dispatch: Alt.Dispatcher[Schema, U]): F[U]
   def biject[A, B](schema: Schema[A], bijection: Bijection[A, B]): F[B]
   def refine[A, B](schema: Schema[A], refinement: Refinement[A, B]): F[B]
-  def lazily[A](suspend: Lazy[Schema[A]]): F[A]
-  def nullable[A](schema: Schema[A]): F[Option[A]]
+  def lazily[A](shapeId: ShapeId, hints: Hints, suspend: Lazy[Schema[A]]): F[A]
```
Comment (Baccata) [inline modules/core/src/smithy4s/schema/SchemaVisitor.scala:32]: Nope, I'm veto-ing this change (same for nullable). If this change is required for the feature to work, I'll veto the feature.

**Review by Baccata — CHANGES_REQUESTED** [review]

**Comment by Baccata** [comment]
Closing this as well as the originating issue : considering the requirement for compiling schemas is `Schema ~> F`, which can take many forms, and considering this feature is not helping the implementation of the OOTB integrations, I don't think it's required at all. 

If you come up with a pattern in userland that you deem really useful, I'm happy to take a look and potentially integrate it, but right now this feels like a "it may come in handy" more than anything.

**Pull request closed without merging**